### PR TITLE
CMake overhaul

### PR DESCRIPTION
**Pull request by Malacath-92**
- Not all changes are backwards compatible because of variable names changing on the CMake interface
- Bump CMake to 3.1
- Merged the options for building shared/static, there is no reason a user would want to build both
- Removed all write to global CMake configure state (aka things as `include_directories` and `CMAKE_CXX_FLAGS`) and replaced with appropriate functions per target
- Moved warnings to a private interface of inicpp, this way users are not forced to enable the same set of warnings (which you probably should fix, there are a lot of warnings avoidable and some probably are actual problems) (also you support exclusively C++17, so [[maybe_unused]] is the correct way to silence warnings for unused symbols)
- Avoid generic variable names in CMake (afaik if a parent scope contains a variable of the same name you may override it)
- Used linking of targets wherever possible instead of duplicating source files
- Configure tests and examples only when explicitly specified
- The changes to the CMake interface may warrant a version bump but that's up to you ofc

Non-CMake changes:
- I work on Windows so I fixed the windows builds when linking against the shared library, see changes to string_utils.h
- To get a better overview of what works and what doesn't I consolidated the GitHub Actions CI into one workflow building tests and examples in Debug, Release, Shared, Static, Windows, MacOS, Ubuntu, Clang, GCC, MSVC whatever combinations are possible. This is up to you if you want to integrate it or not, but if you fix all warnings on those compiler, enable /Wx and -Werror you're gonna start getting a really clean build which your users will appreciate a lot. I'm not sure why the shared configurations for Windows fail, they work on my machine, but I haven't looked closer at the tests, so maybe it's something obvious.
- On that note, when you advertise using CMake you should also start using the cmake cmd line tool for building and installing, that's also reflected in the workflow I wrote.

I tried to make my commits as small as possible so you can easily verify if you agree with the changes, but I may have slipped here or there and made multiple changes in one commit, sorry about that.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+          fi
+        shell: bash
+        if: matrix.os == 'windows-latest'
+
+      - name: Configure
+        run: |
+          mkdir build
+          cd build
+          if [ "$RUNNER_OS" == "Windows" ]; then
+            cmake .. -DINICPP_BUILD_TESTS=ON -DINICPP_BUILD_EXAMPLES=ON -DINICPP_BUILD_SHARED=${{matrix.build_shared}} -T${matrix.msvc_toolset}
```
Comment (Neloop) [inline .github/workflows/build.yml:74]: Variable `matrix.msvc_toolset` enclosed in only one pair of braces, that's why Windows builds are failing

**Comment by Neloop** [comment]
Regarding CI, I'm not really fan of having everything in one workflow. Almost every notable step in your pipeline is having an `if` based on used operating system. There are even two types of ifs. If the build is so different on Windows I would suggest to keep the two workflows for Windows and non-Windows builds.

Thoughts @SemaiCZE ?

**Comment by SemaiCZE** [comment]
Yes, I'd prefer to have two separate workflows without everything `if`ed. First impressions on this PR are good though, but I need to look into it a bit more before merging. Thanks for your work

File: .github/workflows/build.yml (hunk 1 above)
Comment (Malacath-92) [reply .github/workflows/build.yml:74]: Oh lol, I just added that after making the PR, previously it was just running with cl.exe on both the configs and you can see there that the runs that build a shared library fail and the runs that build a static library succeed.

**Comment by Malacath-92** [comment]
Ye, all up to you, as I said I just preferred it to verify for myself if things are working  👍 
Personally I prefer to have a few if statements and get all runs in one workflow, rather than having a big chunk of duplicate setup in two or more workflows. And re the two types of if statements, that's not even necessary, I just derped there, too tired in the morning  😄

File: .github/workflows/build.yml (hunk 1 above)
Comment (Malacath-92) [reply .github/workflows/build.yml]: Fixed in latest commit

**Comment by Malacath-92** [comment]
I split Windows and Non-Windows workflows. But tbh if you want the granularity you may also want split the Linux and MacOS builds so you can have separate badges for those, giving users a better view of what platforms are stable. But I think you can take it from here. But ping me if there's anything in the PR you want some clarity on.

**Comment by SemaiCZE** [comment]
@Malacath-92 I tried to fix the Mac build, but with no success yet. Can you please take a look into thw failing Windows builds? I think that it's something like https://stackoverflow.com/questions/59376924/ctest-cmake-mingw-executables-build-but-fail-to-run-because-fresh-dll-is-n and since you're using Windows you can easily test that. Thanks.

**Comment by Malacath-92** [comment]
I'm not super happy with this solution, but I guess that's what I get for developing on windows  😅

**Pull request merged**